hack/install-tools.sh

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env bash
+
+#######################################
+# This script installs Go programs which defined in hack/tools.go.
+#######################################
+
+set -euo pipefail
+
+#######################################
+# Get imports in a given Go source file
+# Arguments:
+#   A path to the Go source code.
+# Outputs:
+#  Writes space-separated imports to stdout.
+#######################################
+function get_imports() {
+  local src="$1"
+
+  # Listing imports in the given source file
+  local imports
+  imports=$(go list -e -f '{{.Imports}}' "${src}")
+
+  # Trim brackets from the output
+  imports="${imports/[/}"
+  imports="${imports/]/}"
+
+  echo "${imports}"
+}
+
+#######################################
+# Get module version which specified in go.mod of the given package.
+# Arguments:
+#   A name of the package.
+# Outputs:
+#   Writes version of the given package to stdout.
+#######################################
+function get_module_version() {
+  local package="$1"
+
+  # Output of `go list -f '{{.Module}}' "${package}"` is formatted in "<package> <version>",
+  # so treat output as an array and print the second element - version.
+  IFS=" " read -r -a module <<< "$(go list -f '{{.Module}}' "${package}")"
+  echo "${module[1]}"
+}
+
+function main() {
+  local workdir
+  workdir=$(dirname "$0")
+
+  local programs
+  programs=$(get_imports "${workdir}/tools.go")
+
+  for program in ${programs}; do
+    version=$(get_module_version "${program}")
+
+    echo "Installing ${program}@${version}..."
+    go install "${program}@${version}"
```

Comment:
By no means a blocker, but an observation as I have been diving into the build/devx related parts of Kargo the past day. :cherries: 

---

When you are working on multiple Kubernetes related projects, `controller-gen` tends to be on different versions across these projects quite a lot. In addition, assuming that e.g. a contributor does not already have _some_ version of a tool on their system (and us overwriting it) is quite invasive. 

Given this, I am wondering if there have been chats about collecting dependency tools in a directory relative to the project root, e.g. `./hack/bin`. Instead of globally installing them on a developer their system.

A common trick to pull this off is to set `GOBIN` to a newly created temporary directory before running `go install`, and to then copy any binaries from there. See for example: https://github.com/DataDog/datadog-operator/blob/a4cf7a31e463b69dce523f9ef41aab2aaa5c1421/Makefile#L83-L108

Reply:
He opened #1525, so let's count this as resolved.